lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+struct ResetNode {
+  /// The value in the IR that corresponds to this node.
+  FieldRef value;
+  /// The type of the value.
+  FIRRTLType type;
+  /// The reset net this node belongs to.
+  ResetNet *net = nullptr;
+};
+
+/// A global view of all reset networks in a design.
+///
+/// A `ResetMap` associates values in a design with a `ResetNet`. It does this
+/// by creating a `ResetNode` for every value (represented by a `FieldRef`),
+/// which holds a pointer to the reset network the value belongs to. The core
+/// functionality is the `add` function, which stores a connection between two
+/// values in the map. Doing so associates a "source" as the value being driven
+/// onto a "destination" port or wire, and combines the `ResetNet` the values
+/// belong to into a single one. This transitively establishes a `ResetNet` as
+/// the collection of all values in the IR that must have the same reset type,
+/// since they are connected together (through instance ports or connect ops).
+struct ResetMap {
+  ~ResetMap() { clear(); }
+
+  void clear();
+  void add(FieldRef dst, FIRRTLType dstType, FieldRef src, FIRRTLType srcType,
+           Location loc);
+  ResetNode *getNode(FieldRef value, FIRRTLType type);
+  ResetNet *createNet();
+  void abandonNet(ResetNet *net);
+
+  /// Iterate the reset nets in this map.
+  auto begin() { return nets.begin(); }
+  auto end() { return nets.end(); }
+  auto getNets() { return llvm::make_range(begin(), end()); }
+
+private:
+  /// The allocator for reset nodes and nets.
+  BumpPtrAllocator allocator;
+
+  /// A mapping from signals to a corresponding node in a reset network.
+  DenseMap<FieldRef, ResetNode *> nodes;
+
+  /// A list of used reset networks, unused networks, and allocated networks.
+  SmallSetVector<ResetNet *, 8> nets;
+  SmallVector<ResetNet *> unusedNets, allocatedNets;
+};
+} // namespace
+
+template <typename T>
+static T &operator<<(T &os, const ResetKind &kind) {
+  switch (kind) {
+  case ResetKind::Uninferred:
+    return os << "<uninferred>";
+  case ResetKind::Async:
+    return os << "async";
+  case ResetKind::Sync:
+    return os << "sync";
+  }
+}
+
+/// Clear the nodes and drives within the net.
+void ResetNet::clear() {
+  nodes.clear();
+  drives.clear();
+  kind = ResetKind::Uninferred;
+}
+
+/// Determine a good location for this reset network to report to the user. A
+/// reset network is just a bag of IR values and associated connects, so it has
+/// no location per se. However for the sake of diagnostics, we can report a
+/// port or wire somewhere at the top of the network to the user.
+FieldRef ResetNet::guessRoot() {
+  // Count the drives targeting each node.
+  SmallDenseMap<ResetNode *, unsigned> nodeIndex;
+  SmallVector<unsigned> driveCounts(nodes.size(), 0);
+  for (auto it : llvm::enumerate(nodes))
+    nodeIndex.insert({it.value(), it.index()});
+  for (auto &drive : drives)
+    ++driveCounts[nodeIndex[drive.dst]];
```

Comment:
Why indirect through nodeIndex?  Why not just have a `DenseMap<Node*, unsigned>` where the value of the map is the count?

Reply:
This was definitely the most convoluted way to do things. Mental sanity is now restored and this is much simpler.